### PR TITLE
Npc interaction integration

### PR DESCRIPTION
**Pull request by smlendzian**
this is the code of 3 different people (one of which was an entire Sprint's worth and another was almost an entire Sprint's worth), so good luck and thank you very much to whichever of you reviews this

### CODE REVIEW

File: Assets/Scripts/UIAndMenus/UILookAtCam.cs [hunk 1]
```diff
@@ -4,8 +4,9 @@
 
 public class UILookAtCam : MonoBehaviour
 {
-    void LateUpdate()
+    private void LateUpdate()
     {
-        transform.LookAt(transform.position + Camera.main.transform.rotation * Vector3.forward, Camera.main.transform.rotation * Vector3.up);
+        transform.rotation = Quaternion.LookRotation(transform.position - Camera.main.transform.position);
```
Comment (cadenceblorbo) [inline Assets/Scripts/UIAndMenus/UILookAtCam.cs:9]: better behavior would be:
`transform.rotation = Camera.main.transform.rotation;`

File: Assets/test [hunk 2]
```diff
@@ -0,0 +1 @@
+yuh:
```
Comment (cadenceblorbo) [inline Assets/test:1]: why is this here? should probably remove

(yuh)

File: Assets/Scripts/Dialogue System/Dialogue Editor/DialogueGraphView.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+using UnityEditor;
+using UnityEditor.Experimental.GraphView;
+using UnityEditor.UIElements;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+
+public class DialogueGraphView : GraphView
+{
+    public readonly Vector2 defaultNodeSize = new Vector2(150,200);
```
Comment (cadenceblorbo) [inline Assets/Scripts/Dialogue System/Dialogue Editor/DialogueGraphView.cs:14]: This value does not seem to actually be respected, changing it does nothing to the node sizes

File: Assets/Scripts/Dialogue System/Dialogue Editor/GraphSaveUtility.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                });
+            }
+
+        //create folder if necessary
+        if (!AssetDatabase.IsValidFolder("Assets/Resources")) {
+            AssetDatabase.CreateFolder("Assets", "Resources");
+            }
+        //save graph
+        AssetDatabase.CreateAsset(dialogueContainer, $"Assets/Resources/{fileName}.asset");
+        AssetDatabase.SaveAssets();
```
Comment (cadenceblorbo) [inline Assets/Scripts/Dialogue System/Dialogue Editor/GraphSaveUtility.cs:56]: might be a good idea to refactor save location to a single string at top of file that gets referenced to make changing it easier

File: Assets/Scripts/Dialogue System/RunTime/DialogueNodeData.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[Serializable]
+public class DialogueNodeData
+{
+    public string NodeGuID;
+    public string DialogueText;
+    public Vector2 position;
```
Comment (cadenceblorbo) [inline Assets/Scripts/Dialogue System/RunTime/DialogueNodeData.cs:11]: The base data storage for this and NodeLinkData is set up in a less than ideal way. Ideally, nodes should be able to access their children/parents through references, rather than searching a list by GUID.

File: Assets/Scripts/DialogueBox.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    [SerializeField] GameObject dialogueBox;
+    public void showDialogueBox() {
+        dialogueBox.SetActive(true);
+        }
+    public void hideDialogueBox() {
+        dialogueBox.SetActive(false);
+        }
+    
+}
+*/
```
Comment (cadenceblorbo) [inline Assets/Scripts/DialogueBox.cs:18]: Does nothing, remove

File: Assets/Scripts/InventoryAndItemsScripts/AreaThatChecksInventory.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    [RequireComponent(typeof(InventoryChecker))]
+    [RequireComponent(typeof(Collider))]
+    public class AreaThatChecksInventory : MonoBehaviour
+    {
+        public event Action<GameObject> FailedInventoryCheck;
+        public event Action<GameObject> PassedInventoryCheck;
+
+        private void OnTriggerStay(Collider other)
+        {
+            if (other.gameObject.tag == "Player")
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/AreaThatChecksInventory.cs:17]: invert check to reduce nesting
`if (other.gameObject.tag != "Player"){
    return;l
}`

File: Assets/Scripts/InventoryAndItemsScripts/InventoryChecker.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        [SerializeField] List<ItemData> mustHaveEquippedAtLeastOneOfTheseItems;
+
+        /// <summary>
+        /// Checks the PlayerInventory.
+        /// Returns true if the PlayerInventory meets the criteria
+        /// in SerializedFields for this 
+        /// </summary>
+        public bool InventoryStateMeetsCriteria()
+        {
+            if (this.mustHaveEquipped != null &&
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/InventoryChecker.cs:39]: != null is fine in this case, in C# it is better practice to use <var> is not null for more expected behavior
https://stackoverflow.com/questions/69267179/what-is-the-difference-between-is-not-null-and-null
not a necessary change here but is a good practice to get into

File: Assets/Scripts/InventoryAndItemsScripts/InventoryChecker.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            }
+
+            return true;
+        }
+
+        private bool CheckMustHaveItems()
+        {
+            foreach (InventoryItem neededItem in this.mustHaveTheseItems)
+            {
+                if (PlayerInventory.Instance.ItemDictionary.TryGetValue(neededItem.Data, out InventoryItem inventoryItem))
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/InventoryChecker.cs:73]: invert statement and return false to reduce nesting

File: Assets/Scripts/InventoryAndItemsScripts/Items/Collectables/InteractCollectable.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        }
+
+        void Awake()
+        {
+            this.willDestroyWhenDoneCollecting = this.destroyOnCollect;
+            this.destroyOnCollect = false;
+
+            this.interactableOverlay = this.GetComponentInChildren<InteractableOverlayHelper>();
+            if (this.interactableOverlay == null)
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Collectables/InteractCollectable.cs:36]: this.interactableOverlay is null is generally better practice, again, probably not necessary

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/AttachInteractorToPlayer.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            this.playerInteractor = PlayerInteractor.Instance;
+        }*/
+
+        /// <summary>
+        /// Only use this Property after Awake()
+        /// </summary>
+        //public PlayerInteractor PlayerInteractor { get { return this.playerInteractor; } }
+    }
+}
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/AttachInteractorToPlayer.cs:30]: does nothing, remove

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/DialogueInteraction/DialogueHandler.cs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+                buttonPosition += this.buttonDisplacement;
+            }
+        }
+
+        private void CreateButton(Vector2 buttonPosition, NodeLinkData childNode)
+        {
+            GameObject newButton = Instantiate(this.buttonPrefab, buttonPosition, Quaternion.identity, this.parentCanvas.transform);
+            
+            Button newButtonButton = newButton.GetComponentInChildren<Button>();
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/DialogueInteraction/DialogueHandler.cs:65]: would rename to something like buttonComponent

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/DialogueInteraction/DialogueHandler.cs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+        //private NodeLinkData currentNodeLinkData;
+        private DialogueNodeData currentDialogueNodeData;
+
+        private List<GameObject> currentButtons = new List<GameObject>();
+
+        void Awake()
+        {
+            //Get the first node link, which has the PortName "Next"
+            NodeLinkData firstNodeLink = this.dialogueContainer.NodeLinks.First(node => node.PortName == "Next");
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/DialogueInteraction/DialogueHandler.cs:28]: searching unideal, already discussed

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            {
+                OnInInteractionRadius(mayInteract);
+            }
+        }
+
+        void Interact(CallbackContext c)
+        {
+            if (this.interactableCurrentlyInRange == null || !this.mayInteractWithCurrentInteractable)
+            {
+                Debug.LogError("Cannot Interact.");
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs:73]: change to debug.log

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+                OnLeaveInteractionRadius();
+            }
+        }
+
+        //May remove this if it causes performance issues
+        public void CheckWhetherCanStillInteract(Collider playerCollider)
+        {
+            if (this.interactableCurrentlyInRange == null)
+            {
+                Debug.LogError("Interactable not currently in range.");
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs:103]: change to debug.log

File: Assets/Scripts/Dialogue System/Dialogue Editor/DialogueGraphView.cs (hunk 3 above)
Comment (smlendzian) [reply Assets/Scripts/Dialogue System/Dialogue Editor/DialogueGraphView.cs:14]: not changing now

File: Assets/Scripts/Dialogue System/RunTime/DialogueNodeData.cs (hunk 5 above)
Comment (smlendzian) [reply Assets/Scripts/Dialogue System/RunTime/DialogueNodeData.cs:11]: not changing now

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        /// <summary>
+        /// Triggered when a PlayerInteractor that was previously not colliding
+        /// with an InteractionCollider begins colliding with one
+        /// (takes in a bool that is true when the criteria to interact has been met)
+        /// </summary>
+        public event Action<bool> OnInInteractionRadius;
+
+        public event Action OnLeaveInteractionRadius;
+
+        public event Action<bool> CanInteractUpdate;
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs:49]: Are these for later use? As of now I don't see where they are ever assigned/called.

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/UIInteractable.cs [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+                //Pause game (should later replace with game's pause system)
+                Time.timeScale = 0;
+            }
+            else
+            {
+                Destroy(this.instantiatedUIPrefab);
+
+                //UnPause game (should later replace with game's pause system)
+                Time.timeScale = 1;
+            }
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/UIInteractable.cs:56]: For a toggle like this, should probably enable/disable the same GameObject instead of making a new one each time

File: Assets/Scripts/InventoryAndItemsScripts/ScriptsForTestingInventory/TestDoor.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+            InventoryChecker checker = this.GetComponent<InventoryChecker>();
+
+            if (checker.InventoryStateMeetsCriteria())
+            {
+                this.GetComponent<Collider>().isTrigger = true;
+            }
+            else
+            {
+                this.GetComponent<Collider>().isTrigger = false;
+            }
```
Comment (cadenceblorbo) [inline Assets/Scripts/InventoryAndItemsScripts/ScriptsForTestingInventory/TestDoor.cs:36]: could simplify to this.GetComponent<Collider>().isTrigger =checker.InventoryStateMeetsCriteria(), stylistic choice

File: Assets/DialogueBox.cs [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+public class DialogueBox : MonoBehaviour
+{
+    [SerializeField] GameObject dialogueBox;
+    public void showDialogueBox() {
+        dialogueBox.SetActive(true);
+        }
+    public void hideDialogueBox() {
+        dialogueBox.SetActive(false);
+        }
+}
```
Comment (cadenceblorbo) [inline Assets/DialogueBox.cs:14]: Unused, but if it's supposed to be used for the game object it's attached to, you should use this.SetActive() instead of passing itself trough a field.

**Review by cadenceblorbo — COMMENTED** [review]
This is a big request, so may be a good idea to get another pair of eyes on it.

Overall thoughts:

-Dialogue graph system is a good candidate for refactor. Most glaring issue is the way neighbors are interacted with, but could also be improved through visual changes and splitting behavior into shorter methods. I'd say anything that impacts people using the system is a good first thing to clean up.

-Interactable system is very intuitive, clean, and flexible

-Interaction UI still needs some polish, can come later

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/DialogueInteraction/DialogueHandler.cs (hunk 13 above)
Comment (smlendzian) [reply Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/DialogueInteraction/DialogueHandler.cs:28]: not changing now

File: Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs (hunk 16 above)
Comment (smlendzian) [reply Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs:49]: they were used a bit ago when I had a completely different test script calling them. At the moment I'm keeping them later in case they may still be used so I don't need to re-code them, but if it's an issue I can remove them.

Comment (cadenceblorbo) [reply Assets/Scripts/InventoryAndItemsScripts/Items/Interactables/PlayerInteractor.cs:49]: I think good to keep for now!

File: Assets/Scripts/InventoryAndItemsScripts/ScriptsForTestingInventory/TestDoor.cs (hunk 18 above)
Comment (smlendzian) [reply Assets/Scripts/InventoryAndItemsScripts/ScriptsForTestingInventory/TestDoor.cs:36]: I'm going to leave this as is for now since we will need to update how this works anyway to work with A. the Unity CharacterController collision B. however we want to animate doors
